### PR TITLE
Return all leases before OnResuming is called for a resuming screen

### PR DESCRIPTION
**Pull request by nyquillerium**
Closes https://github.com/ppy/osu/issues/4377

Tested locally against osu

### CODE REVIEW

File: osu.Framework.Tests/Visual/TestCaseUserInterface/TestCaseScreenStack.cs [hunk 1]
```diff
@@ -295,12 +310,26 @@ private class TestScreen : Screen
...
+            private readonly bool shouldTakeOutLease;
+
             internal override void UnbindAllBindables()
             {
                 base.UnbindAllBindables();
                 OnUnbind?.Invoke();
             }
 
+            public TestScreen(bool shouldTakeOutLease = false)
+            {
```
Comment (peppy) [inline osu.Framework.Tests/Visual/TestCaseUserInterface/TestCaseScreenStack.cs]: is this necessary?

Comment (nyquillerium) [reply osu.Framework.Tests/Visual/TestCaseUserInterface/TestCaseScreenStack.cs]: not necessary

File: osu.Framework.Tests/Visual/TestCaseUserInterface/TestCaseScreenStack.cs [hunk 2]
```diff
@@ -295,12 +310,25 @@ private class TestScreen : Screen
             public override bool HandleNonPositionalInput => true;
             public Action OnUnbind;
 
+            public bool IsBindablesReturned;
```
Comment (peppy) [inline osu.Framework.Tests/Visual/TestCaseUserInterface/TestCaseScreenStack.cs]: bindables?

File: osu.Framework/Screens/ScreenStack.cs [hunk 3]
```diff
@@ -206,6 +208,16 @@ private void exitFrom([CanBeNull] IScreen source, Action onExiting = null)
...
 
+        /// <summary>
+        /// Unbind and return leases for all <see cref="Bindable{T}"/>s managed by the exiting screen.
+        /// </summary>
+        /// <remarks>
+        /// While all bindables will eventually be cleaned up by disposal logic, this is too late as
+        /// leases could potentially be in a leased state during exiting transitions.
+        /// This method should be called after exiting is confirmed to ensure a correct leased state before <see cref="IScreen.OnResuming"/>.
+        /// </remarks>
+        private void onExited(IScreen prev, IScreen next) => (prev as Screen)?.UnbindAllBindables();
```
Comment (peppy) [inline osu.Framework/Screens/ScreenStack.cs]: this cast is a bit dangerous since `IScreen`s will not get this applied to them. @smoogipoo thoughts? Might be better to cast to `CompositeDrawable` instead?

File: osu.Framework.Tests/Visual/TestCaseUserInterface/TestCaseScreenStack.cs [hunk 4]
```diff
@@ -410,6 +444,10 @@ public override void OnResuming(IScreen last)
             {
                 ResumedFrom = last;
 
+                // Check if the last screen's bindables have since been returned
+                if (!((TestScreen)last).leasedCopy?.Disabled ?? false)
```
Comment (peppy) [inline osu.Framework.Tests/Visual/TestCaseUserInterface/TestCaseScreenStack.cs]: this condition is really hard to parse. why not check `DummyBindable.Disabled` instead, now that you're correctly binding down the stack?

you can remove IsBindableRetuned and just check `DummyBindable.Disabled` directly.
